libhypatia/src/segment.rs

```diff
@@ -0,0 +1,25 @@
+/// Call this macro from within each segment to include all the per-segment boilerplate.
+#[macro_export]
+macro_rules! define_segment {
+    () => {
+        mod _segment {
+            use core::panic::PanicInfo;
+
+            #[cfg(not(test))]
+            #[panic_handler]
+            pub extern "C" fn panic(_info: &PanicInfo) -> ! {
+                #[allow(clippy::empty_loop)]
+                loop {}
+            }
+
+            #[cfg(not(test))]
+            #[lang = "eh_personality"]
+            extern "C" fn eh_personality() {}
+
+            #[cfg_attr(not(test), no_mangle)]
+            pub extern "C" fn main() {
```

Comment:
I see that this is being used for tasks as well as segments, which is good, but in tasks, `main` ought to be in `main.rs`. Perhaps we can have a segment-specific macro that includes this boilerplate, and a variant that defines boilerplate slightly differently for tasks?

Reply:
Makes sense.

Reflecting on how I messed this up, I think the fact it was all open coded made it hard for me to see the differences.  Your callouts here really help me see the intent, but I think it'll be clearer by the time this pr is done :)

I do find the name "segment" confusing, as as best I can tell, it looks like a shared stateful library without threads of its own?  I find it confusing because "segment" seems overloaded from the memory span concept, which tasks also have/own/are.

Also unclear to me right now is intended visibility of the memory backing tasks within a taskset, do tasks have distinct memory spans within a taskset?

Reply:
> Makes sense.
> 
> Reflecting on how I messed this up,

I wouldn't put it that way. :-)

> I think the fact it was all open coded made it hard for me to see the differences. Your callouts here really help me see the intent, but I think it'll be clearer by the time this pr is done :)

Yeah. Once the PR is in, it will be a major improvement.

Something else I think this highlights is the need to focus more on documentation.  Some of these things are covered in the HDPs, but there's a lot that's still just kind of in my head (and Cory's et al) that needs to get written down.

> I do find the name "segment" confusing, as as best I can tell, it looks like a shared stateful library without threads of its own? I find it confusing because "segment" seems overloaded from the memory span concept, which tasks also have/own/are.

That name is lifted directly from Multics, where segments are named, shared, and have their own code _and_ data. They're a bit of a degenerate case in Hypatia (we're aiming for simplicity; full on Multics-style segmentation would be cool but is way beyond the scope of a simple hypervisor), but it really is analogous in a lot of ways. HDP 0009 talks about this a bit: https://github.com/hypatia-hypervisor/hypatia/tree/main/docs/hdp/0009

> Also unclear to me right now is intended visibility of the memory backing tasks within a taskset, do tasks have distinct memory spans within a taskset?

Possibly, but I prefer to look at it slightly differently: tasks have their own address spaces, which may link in segments (and some segments are _always_ in the address space, for instance the global and node-/cpu-local segments). A task set is just an identifier that indicates that a set of tasks are logically grouped, similar to a process group in Unix/Linux/POSIX. Task sets carry the additional property that members have the right to walk the page tables of the task set leader.

The canonical example is a virtual machine, which is represented as a task set consisting of a single VM control task and one or more VCPU tasks. These (will) set up a shared virtual memory region between them (using the "task set area" in the memory map (https://github.com/hypatia-hypervisor/hypatia/blob/main/docs/hdp/0003/mem.txt). The VM control task is the first to start, and it sets up the shared region and starts the vcpu tasks; those look at the control task's page tables to find where the shared region is mapped and duplicate those mappings in their own address spaces.

This is described at a high level in HDP 0010: https://github.com/hypatia-hypervisor/hypatia/tree/main/docs/hdp/0010

All of this documentation needs a good sweep through and elaboration.

Reply:
Okay, this helps a bit.  I had read HDP9 but it didn't really help me understand the concept of it vs Task.  If I'm understanding you, a Task is a single thread operating in its own address space and we share code+state into that task at runtime.  Other than the association to a taskset, they are otherwise free running, but are still cooperatively scheduled.

Does that sound about right?

Related but also not: Have you guys put some thought into the marshaling process for this segment gate mechanism?  How about timers / asynchronicity / message loops generally?

Reply:
That's close. A task is very like a process and owns its own address space, except that it is cooperatively scheduled and may yield explicitly in a thread-like manner.  It is like a thread in that way, but unlike a thread in the sense of owning its own (unshared) address space. We chose the name "task" because it was neither a process (in the Unix sense, with all of the properties that implies) nor a thread.

A segment is kind of like a shared library, except that it owns its own data, BSS, etc that are also shared.

There's no real marshaling across the segment call boundary, per se; calling into a segment looks an awful lot like a function call, and its anticipated that you're on the task's stack when you do it: conceptually, it's as if you are trampolining through an array of function pointers. That array is in some known location in the portion of the address space dedicated to that segment, and entries have fixed offsets. E.g., transfer vector 0 is foo, vector 1 is bar, and so on.

Tasks should be prepared to accept a small subset of interrupts (we reserved the top 16 or something for the hypervisor), so that we can signal between cores. The basic messaging flow is that each task owns a mailbox that can be mapped somewhere by a sender. The sender allocates a slot in the mailbox, writes the message, and sets a "pending" bit for the slot and tries to send a doorbell interrupt. When the task runs, it checks its mailbox for pending messages, acts on them accordingly, writes the reply into mailbox (each slot contains space for both the message _and_ reply), marks the message replied and possibly sends a doorbell.

An interesting property of the system is that the various tasks and segments and so on are statically provided with the system and "trusted" in that sense, so we can get away with doing things you just couldn't rely on in a general purpose OS (like cooperative scheduling). A task may take a "reschedule yourself" interrupt, but it's still its responsibility to yield in a timely manner. Since we wrote it, we can return control to it trusting it to do the right thing; this gives it a chance to put itself into a known state before it yields, which ties into timers and so on. The task knows how long it is going to run (or, rather, the scheduler can tell it) and it can arrange to send itself a timer interrupt, thus giving many of the properties of preemptive scheduling in a cooperative framework.

Gosh I really need to write this stuff down.

Reply:
Hmm.  What you describe sounds like a OG Unix process, like before there were threads.  The main difference here seems to be the focus on cooperative scheduling.

Re marshalling I guess I was looking for any such tables or examples of how this is expected to work. Sounds like the interface for each segment in the end is a big old manual vtable that is manually populated by the segment (at init I presume?) and that code gets dispatched through?  I guess I'm wondering how the caller side stubs for the segment call table is done / can be made simple.

I see the simplicity (and potential data gating benefits) in the interrupt model, but wonder about exposing them to segments directly. It would be much cleaner for instance if say interrupts were solely handled at the runtime/ipc level and code running in segments never had to deal with reentrancy nor edge-vs-level signalling issues brought about the interrupt/signal pattern. Something that provided respondable mpsc channels, that maybe could be tied into async's poll functionality seems much more preferable over ever having to deal with interrupt gates.

For the benefit of discussion, I think it would be good to work through the general extension of the 'reschedule yourself' interrupt, which I think is just the "timer service".  In this case, what do we propose the flow of execution look like?  I can only assume that this would end up living in one of the segments as a global service?  Would the forced intention be to have the callbacks always come in through a call gate, or does it make sense to share a callback function pointer to the timer service in this case (I'm not sure it does?).

I think mechanically, as a developer, I'd still want the latter, even if the underlying architecture for this task/segment construction is doing the former for reasons.

Glad I can force you to write words.  Lots of this will get copy and pasted / revised I'm sure :)

Reply:
Hey Mike, just a quick response...I apologize for the brevity but we're leaving for a week for vacation in a few hours but I didn't want to blow off your note. We'll be back in a week, and I was thinking it might be worthwhile to try to schedule a meet-and-greet Hypatia meeting, if you'd be interested? I'll ask some G folks and maybe some Oxiders if they'd like to join, as well.

> Hmm. What you describe sounds like a OG Unix process, like before there were threads. The main difference here seems to be the focus on cooperative scheduling.

Kind of. Unix processes have pretty much always shared the entire kernel in the "upper" part of the address space; Hypatia tasks aren't like that precisely.

> Re marshalling I guess I was looking for any such tables or examples of how this is expected to work. Sounds like the interface for each segment in the end is a big old manual vtable

(Hopefully it's really small. :-))

> that is manually populated by the segment (at init I presume?) and that code gets dispatched through? I guess I'm wondering how the caller side stubs for the segment call table is done / can be made simple.

Yeah. It's sort of like manual PLT. Glenn asked a while back, "why not just use actual shared libraries?" and the reason is so that we have a tightly controlled interface into segments.

The vision is that segment use will be pretty limited and infrequent; most of the time, a VCPU task will be running (in guest context, one hopes). Occasionally it'll pop out and the back in; every so often we'll context switch between VCPU tasks and maybe run the VM control task once in a while.

> I see the simplicity (and potential data gating benefits) in the interrupt model, but wonder about exposing them to segments directly. It would be much cleaner for instance if say interrupts were solely handled at the runtime/ipc level and code running in segments never had to deal with reentrancy nor edge-vs-level signalling issues brought about the interrupt/signal pattern. Something that provided respondable mpsc channels, that maybe could be tied into async's poll functionality seems much more preferable over ever having to deal with interrupt gates.

Well, segments don't have to care so much; handling interrupts is really the responsibility of a task, and the set of interrupts they have to deal with is small: really just a doorbell, go-schedule-yourself, and TLB shoot-downs. The CSP-style things would imply more memory sharing than we were particularly comfortable with, and we ran into corner cases with flow control that we were concerned about with the queued, in-order message model.

> For the benefit of discussion, I think it would be good to work through the general extension of the 'reschedule yourself' interrupt, which I think is just the "timer service".

Not just. For example, the root VM may start a new virtual machine and recreate the per-core schedules; it would notify the cores involved via sending them a scheduler IPI.

> In this case, what do we propose the flow of execution look like?

I think there are two cases here: either a timer goes off, or another core requests a reschedule. In either case, the flow looks something like:

1. A task is scheduled on a CPU (note: I've been sloppy with terminology and using core/thread interchangeably, but for discussion purposes, I really mean a hyperthread).
2. That task owns and presents to the rest of the system an entry point that is the, "I got a scheduler interrupt". The details of how this works are a bit superfluous, but generally there's a per-node IDT that points to a per-CPU dispatch table and part of the task preparing itself to run involves writing an entry into that table pointing to the task's interrupt handler; the general interrupt code runs, preserves state, and jumps to the appropriate entry in the dispatch table, thus invoking the task's handler. I'm glossing over many details for brevity, of course, but one may assume there's defensive code to ensure that the dispatch vector is valid and so forth.
3. The task runs; generally it's going to start executing in the guest, checking its mailbox and posted interrupt descriptor on the way in and possibly handling messages and arranging for interrupt delivery in the guest and so on (lookin' at you, NMIs).
4. At some point the vcpu task pops out of the guest, either due to receipt of an external interrupt (timer, IPI, whatever), expiration of e.g. the VMX preemption timer (or whatever AMD equivalent exists), arrival of a message in the task's mailbox, or guest action. The vcpu will figure out why it was yanked out of the guest and act accordingly.
5. Suppose the exit reason was the expiration of the task's time quanta, or an external request to reschedule; the vcpu will do whatever work it needs to prepare itself to be preempted (preserve state, etc), and then disable interrupts and jump into the scheduler.
6. The scheduler will figure out what to do: either context switch to another task, or resume execution of the current task.
7. And we return to step 1.

> I can only assume that this would end up living in one of the segments as a global service? Would the forced intention be to have the callbacks always come in through a call gate, or does it make sense to share a callback function pointer to the timer service in this case (I'm not sure it does?).

Specific handling for preemption of this type lives in each task. This is one of the nice properties of the model: since the set of tasks is small and all bundled with the system, we can tailor the response to each task and "trust" it to do the right thing.

I think the note about the function pointer is largely accurate; the task is responsible for depositing a handler for its preemption routine in the CPU-local dispatch table when scheduled.

Obviously there'll be some bits that are common and shared between different task types; I suspect that code will live in a normal library and just be linked statically into the text of each task binary. I believe it will be sufficiently small that the overhead will be minimal.

> I think mechanically, as a developer, I'd still want the latter, even if the underlying architecture for this task/segment construction is doing the former for reasons.

I think in practice that's the way it'll work out, but in this model, things are pretty explicit.

> Glad I can force you to write words. Lots of this will get copy and pasted / revised I'm sure :)

Indeed it will! Thanks, it's a good forcing function. :-)

Reply:
> it might be worthwhile to try to schedule a meet-and-greet Hypatia meeting, if you'd be interested?

Sure let me know when you are thinking.

> Kind of. Unix processes have pretty much always shared the entire kernel in the "upper" part of the address space; Hypatia tasks aren't like that precisely.

That's sounds like maybe a privileged segment, mapped in for performance reasons? Not really necessary for the abstraction however.

> (Hopefully it's really small. :-))

LOL.

LOL at the "hope" above, but also LOL at myself for the pages I wrote in response (and then deleted) about the width of this design space and potential requirements for cross-segment lifetime management across segments here.

Basically I'm having a hard time seeing how we get away from implementing smt akin to references/handles/descriptors across segments/tasks. As in, smt that is managed such that cleanup can happen on task failure (sync or async to failure). Maybe I'm getting ahead of myself here.

> Well, segments don't have to care so much; handling interrupts is really the responsibility of a task [...]

A vtable that is required for code entry into a segment looks very similar to an interrupt / gate table / syscall table to me.  Sure, extras can exist, like potentially changing hardware mappings or pivoting stacks, but also I sorta expect this to maybe happen upon entry (and maybe exit) to code within a segment, no? Seems callee dependent to me.

I'm assuming this because I'm also assuming the desire for gating segment entries, and minimizing them, is to at some level cope with data privilege separation for speculation attack purposes.

> [...]

I think I agree and understand the rest of what you wrote. I think one way to think about the "service" provided by whatever segment provides the timer service, it is best if the shared component do not do much more than expose what is effectively a virtual hardware function of the device in question (maintains minimal state).  By that, I mean that the task sees a thing that looks like virtual normalized timer device exposed in software by the common facility exposed by a segment to tasks (each task gets a single timer service?), and dispatch is library code in the task.  Seems familiar enough, modulo the need to do enter timer code via interrupt *cough* task signal.

Reply:
Ok, I'm back in the country and (mostly) caught up. Sorry for the delayed response. Maybe a meeting in the next couple of weeks?

Reply:
Sure, if you set something up, I'd be happy to join.